### PR TITLE
Bug Fix for bulk setting of URL Aliases

### PR DESCRIPTION
**Pull request by spficklin**
# Bux Fix

Issue #727

## Description

There are two problems when updating URL alias.  
1)  Somehow in the code below the `$bundle_tokens` didn't get used properly in the for loop causing errors when updating.
https://github.com/tripal/tripal/blob/3f134c3030d4b6e8ff834ecef05626ed0eb311ab/tripal/includes/tripal.bulk_update.inc#L19-L30
2)  If the user uses the word `feature` as the first word in the URL it conflicts with Tripal's feature lookup tool that already uses that URL.
3)  In issue #727 it was reported that the URL wouldn't work on the first attempt if manually typing the URL into the browser. I think this may be related to number 2 above, as I can only reproduce the problem in that case. But we will test for it here to be sure.

This PR fixes the code in issue #1 and adds a warning that user's should not use `feature` as the first word.

## Testing?

1) First pull the branch.

2) Go to Admin > Structure > Tripal Content Types and select a content type that is mapped to the Chado feature table (e.g. Gene).   Edit the content type and scroll to the bottom and click the `URL Path Options`   Set the URL to this:
```
feature/[obi__organism,TAXRANK:0000005]/[obi__organism,TAXRANK:0000006]/[rdfs__type]/[data__identifier]
```
Then click the button to `Bulk Update all aliases`   You should get an error notice that the word 'feature' should not be used as the first word in the URL alias string.    

2)  Next, replace the word `feature` with something else (e.g. 'gene') and click the button 'Bulk Upload all aliases'.  it should add a job.

3)  Run the job

4)  To test problem 3 above, do not load a gene page via the GUI, but rather try to type the URL into a browser. If you have sample data from the Tripal tutorial then this URL should work:  http://localhost/blah/Citrus/sinensis/Gene/orange1.1g015632m.g

### CODE REVIEW

**Comment by bradfordcondon** [comment]
word of warning to anyone who tests this: make sure theres no whitespace at the start of the alias (which i got from copy-pasting stephen's instructions in) as the job will submit without a warning and error on run.  I edited @spficklin 's post to remove the whitespace.

I'm able to get the warning and validation to trigger when providing `feature/...`.


Stephen, I get these warnings on my job when providing `gene/[obi__organism,TAXRANK:0000005]/[obi__organism,TAXRANK:0000006]/[rdfs__type]/[data__identifier]` to gene as the URL Alias Pattern.

Sorry I'm not super on top of this issue, but im guessing this isnt what i want?

```
[site http://default] [TRIPAL WARNING] [TRPENTITY] The URL alias, gene///Gene/, already exists for another page. Please ensure the pattern supplied on the <a href="/admin/structure/bio_data/manage/bio_data_8" target="_blank">Gene Edit Page</a> under URL Path options is unique.
[site http://default] [TRIPAL WARNING] [TRPENTITY] The URL alias, gene///Gene/, already exists for another page. Please ensure the pattern supplied on the <a href="/admin/structure/bio_data/manage/bio_data_8" target="_blank">Gene Edit Page</a> under URL Path options is unique.
[site http://default] [TRIPAL WARNING] [TRPENTITY] The URL alias, gene///Gene/, already exists for another page. Please ensure the pattern supplied on the <a href="/admin/structure/bio_data/manage/bio_data_8" target="_blank">Gene Edit Page</a> under URL Path options is unique.
[site http://default] [TRIPAL WARNING] [TRPENTITY] The URL alias, gene///Gene/, already exists for another page. Please ensure the pattern supplied on the <a href="/admin/structure/bio_data/manage/bio_data_8" target="_blank">Gene Edit Page</a> under URL Path options is unique.
[site http://default] [TRIPAL WARNING] [TRPENTITY] The URL alias, gene///Gene/, already exists for another page. Please ensure the pattern supplied on the <a href="/admin/structure/bio_data/manage/bio_data_8" target="_blank">Gene Edit Page</a> under URL Path options is unique.
[site http://default] [TRIPAL WARNING] [TRPENTITY] The URL alias, gene///Gene/, already exists for another page. Please ensure the pattern supplied on the <a href="/admin/structure/bio_data/manage/bio_data_8" target="_blank">Gene Edit Page</a> under URL Path options is unique.
[site http://default] [TRIPAL WARNING] [TRPENTITY] The URL alias, gene///Gene/, already exists for another page. Please ensure the pattern supplied on the <a href="/admin/structure/bio_data/manage/bio_data_8" target="_blank">Gene Edit Page</a> under URL Path options is unique.
8/8 entities have been updated. Memory usage: 34,386,336 bytes.
Done.The URL alias, gene///Gene/, already exists for another page. Please ensure the pattern supplied on the Gene Edit Page under URL [warning]
Path options is unique.
The URL alias, gene///Gene/, already exists for another page. Please ensure the pattern supplied on the Gene Edit Page under URL [warning]
Path options is unique.
The URL alias, gene///Gene/, already exists for another page. Please ensure the pattern supplied on the Gene Edit Page under URL [warning]
Path options is unique.
The URL alias, gene///Gene/, already exists for another page. Please ensure the pattern supplied on the Gene Edit Page under URL [warning]
Path options is unique.
The URL alias, gene///Gene/, already exists for another page. Please ensure the pattern supplied on the Gene Edit Page under URL [warning]
Path options is unique.
The URL alias, gene///Gene/, already exists for another page. Please ensure the pattern supplied on the Gene Edit Page under URL [warning]
Path options is unique.
The URL alias, gene///Gene/, already exists for another page. Please ensure the pattern supplied on the Gene Edit Page under URL [warning]
Path options is unique.
```

**Comment by bradfordcondon** [comment]
stephen says this is because i have orphaned genes.  @bradfordcondon please empty gene bundle, republish, try again.

**Comment by bradfordcondon** [comment]
i deleted the problematic entity.  this is the console feedback:

````
Calling: tripal_update_all_urls_and_titles(bio_data_8, gene/[obi__organism,TAXRANK:0000005]/[obi__organism,TAXRANK:0000006]/[rdfs__type]/[data__identifier], alias)
Array
(
    [0] => Array
        (
            [0] => [obi__organism,TAXRANK:0000005]
            [1] => [obi__organism,TAXRANK:0000006]
            [2] => [rdfs__type]
            [3] => [data__identifier]
        )

)
```

looks like theres a var_dump still in the code?

otherwise functionality confirmed.

File: tripal/includes/tripal.bulk_update.inc
Comment (bradfordcondon) [inline tripal/includes/tripal.bulk_update.inc]: this looks like the debugging variable, lets remove it right @spficklin ?

Comment (bradfordcondon) [inline tripal/includes/tripal.bulk_update.inc]:
```suggestion
```

Comment (spficklin) [reply tripal/includes/tripal.bulk_update.inc]: Oh right!  Good catch.

**Review by laceysanderson — COMMENTED** [review]
Code looks good to me! However, I was able to add spaces in the URL and "  feature/" makes it through your Regex. Now of course no one Should do that but maybe we should add a quick check for it?

**Comment by spficklin** [comment]
I'll add a fix for spaces.

**Comment by spficklin** [comment]
@laceysanderson I have added code to trim out whitespace around the URL alias input by the user. I've tested it and it works.  I also added the missing `drupal_path_alias_whitelist_rebuild()` after setting URL aliases to hopefully fix issue #727.  I've tested it and saw no serious performance issues. It only runs once after all updates of aliases.   I'll leave it up to you to merge this in if you are happy with these changes.

**Review by laceysanderson — APPROVED** [review]
Looks good 👍

**Pull request merged**
